reaktive/src/nativeCommonMain/kotlin/com/badoo/reaktive/scheduler/BufferedExecutor.kt

```diff
@@ -1,50 +1,60 @@
 package com.badoo.reaktive.scheduler
 
-import com.badoo.reaktive.utils.atomic.AtomicReference
-import com.badoo.reaktive.utils.atomic.getAndUpdate
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.utils.RefCounter
+import com.badoo.reaktive.utils.atomic.AtomicBoolean
+import com.badoo.reaktive.utils.lock.Lock
+import com.badoo.reaktive.utils.lock.synchronized
+import com.badoo.reaktive.utils.queue.LinkedFreezableQueue
+import com.badoo.reaktive.utils.queue.Queue
+import com.badoo.reaktive.utils.use
 
 internal actual class BufferedExecutor<in T> actual constructor(
     private val executor: Scheduler.Executor,
     private val onNext: (T) -> Unit
-) {
+) : Disposable {
 
-    private val state = AtomicReference(State<T>())
+    private val lock = Lock()
+    private val refCounter = RefCounter(lock::destroy)
+    private val queue: Queue<T> = LinkedFreezableQueue()
+    private val isDraining = AtomicBoolean()
     private val drainFunction = ::drain
 
+    private val _isDisposed = AtomicBoolean()
+    override val isDisposed: Boolean get() = _isDisposed.value
+
+    override fun dispose() {
+        _isDisposed.value = true
+        refCounter.release()
+    }
+
     actual fun submit(value: T) {
-        val oldState =
-            state.getAndUpdate {
-                it.copy(
-                    queue = it.queue + value,
-                    isDraining = true
-                )
+        refCounter.use {
```

Comment:
Unlike in the jvm BufferedExecutor it is not possible to continue using a disposed executor and will throw an exception here.

Reply:
The `RefCounter` throws an exception only if `release()` is called when the `RefCounter` if destroyed (ref count reached 0). If the `retain()` call returned `false` then the block won't be executed and the `release()` won't be called, and so wan't crash.

Reply:
Right, my mistake 👍 . Mixed the two up.

In that case I think both shouldn't accept input when disposed.